### PR TITLE
Feature/transforms speedup

### PR DESCRIPTION
**Pull request by SamRWest**
Speeds up two slow parts of the `generate_commodity_groups` transform by replacing nested for loops with vectorised groupby calls.

Also adds unit tests for the added functions and small test data files (from austimes) for verification.
Pytest runs and coverage reporting are now integrated into CI as well.

CSV-only regression test results show decent overall speedup:
```
Total runtime: 286.65s (main: 310.07s)
Change in runtime (negative == faster): -23.42s (-7.6%)
Change in correct rows (higher == better): +469 (+0.6%)
Change in additional rows: -12 (-0.0%)
```
But not sure whether the additional row metric increasing or decreasing is good, can you advise?

Update: full GAMS CI run here didn't show any change to correct/additional rows  :)
```
Total runtime: 293.08s (main: 314.38s)
Change in runtime (negative == faster): -21.30s (-6.8%)
Change in correct rows (higher == better): +0 (+0.0%)
Change in additional rows: +0 (+0.0%)
```

### CODE REVIEW

File: xl2times/transforms.py [hunk 1]
```diff
@@ -981,19 +985,11 @@ def generate_commodity_groups(
...
-                i_reg_prc_cset = i_reg_prc & (comm_groups["csets"] == cset)
-                for io in ["IN", "OUT"]:
-                    i_reg_prc_cset_io = i_reg_prc_cset & (comm_groups["io"] == io)
-                    comm_groups.loc[i_reg_prc_cset_io, "commoditygroup"] = sum(
-                        i_reg_prc_cset_io
-                    )
+
+    # Original logic, slow for large tables
+    # _count_comm_group_looped()
+    # Much faster vectorised version
```
Comment (olejandro) [inline xl2times/transforms.py:991]: Should we delete these, since the original is replaced?

File: xl2times/transforms.py [hunk 2]
```diff
@@ -1011,23 +1007,10 @@ def name_comm_group(df):
...
-                for cset in csets_ordered_for_pcg:
-                    i_reg_prc_io_cset = i_reg_prc_io & (comm_groups["csets"] == cset)
-                    df = comm_groups[i_reg_prc_io_cset]
-                    if not df.empty:
-                        comm_groups.loc[i_reg_prc_io_cset, "DefaultVedaPCG"] = True
-                        default_set = True
-                        break
+    # original logic, slow for large tables
+    # comm_groups = pcg_looped(comm_groups, csets_ordered_for_pcg)
+    # vectorised logic, much faster
```
Comment (olejandro) [inline xl2times/transforms.py:1012]: Should we delete these, as well?

File: xl2times/transforms.py [hunk 3]
```diff
@@ -1061,6 +1044,62 @@ def name_comm_group(df):
...
+        comm_groups.groupby(["region", "process", "csets", "io"]).transform("count")
+    )["commoditygroup"]
+    # set comoditygroup to 0 for io rows that aren't IN or OUT
+    comm_groups.loc[~comm_groups["io"].isin(["IN", "OUT"]), "commoditygroup"] = 0
+
+
+def _process_comm_groups_vectorised(
+    comm_groups_test: pd.DataFrame, csets_ordered_for_pcg: list[str]
+) -> pd.DataFrame:
+    """Dets the first commodity group in the list of csets_ordered_for_pcg as the default pcg for each region/process/io combination,
```
Comment (olejandro) [inline xl2times/transforms.py:1066]: Typo: "Sets" instead of "Dets"?

File: xl2times/transforms.py [hunk 4]
```diff
@@ -1061,6 +1044,62 @@ def name_comm_group(df):
...
+
+    comm_groups["commoditygroup"] = (
+        comm_groups.groupby(["region", "process", "csets", "io"]).transform("count")
+    )["commoditygroup"]
+    # set comoditygroup to 0 for io rows that aren't IN or OUT
+    comm_groups.loc[~comm_groups["io"].isin(["IN", "OUT"]), "commoditygroup"] = 0
+
+
+def _process_comm_groups_vectorised(
+    comm_groups_test: pd.DataFrame, csets_ordered_for_pcg: list[str]
```
Comment (olejandro) [inline xl2times/transforms.py:1059]: Do we need to call it `comm_groups_test`?

File: xl2times/transforms.py [hunk 5]
```diff
@@ -1061,6 +1044,62 @@ def name_comm_group(df):
...
+        for io in ["OUT", "IN"]:
+            for cset in csets_ordered_for_pcg:
+                group.loc[
+                    (group["io"] == io) & (group["csets"] == cset), "DefaultVedaPCG"
+                ] = True
+                if group["DefaultVedaPCG"].any():
+                    break
+        return group
+
+    comm_groups_test["DefaultVedaPCG"] = None
```
Comment (olejandro) [inline xl2times/transforms.py:1088]: Can we call it `comm_groups` instead of `comm_groups_test`?

**Comment by olejandro** [comment]
Thanks @SamRWest, looks great! I left some minor comments. I guess, we'll need to merge the benchmark PR first?

**Comment by olejandro** [comment]
> But not sure whether the additional row metric increasing or decreasing is good, can you advise?

It really depends. :-) Some of them could be additional, because casing is not the same as in the benchmarks - doesn't matter for GAMS; some need extra work - e.g. a dimension is inccorrect; some should have not been generated...

**Comment by SamRWest** [comment]
> Thanks @SamRWest, looks great! I left some minor comments. I guess, we'll need to merge the benchmarks PR first?

Yep.  Looks like they're both ready to merge now, just need a couple of approvals. 
But yes, do #184 first please.

**Comment by olejandro** [comment]
@siddharth-krishna please go ahead merging if you don't have any further comments.
It is fun collaborating across the time zones!

File: README.md [hunk 6]
```diff
@@ -94,8 +133,8 @@ python -m pip install --upgrade build
 python -m pip install --upgrade twine
 rm -rf dist
 python -m build
-python -m twine upload dist/*
-```
+python -m twine upload dist/*```
+
```
Comment (siddharth-krishna) [inline README.md:137]: Not sure you meant to make this little change?

File: utils/run_benchmarks.py [hunk 7]
```diff
@@ -310,17 +336,33 @@ def run_all_benchmarks(
...
+    print(f"Total runtime: {our_time:.2f}s (main: {main_time:.2f}s)")
+    print(
+        f"Change in runtime (negative == faster): {runtime_change:+.2f}s ({100*runtime_change/main_time:+.1f}%)"
+    )
+
+    our_correct = df["Correct"].sum()
+    main_correct = df["M Correct"].sum()
+    correct_change = our_correct - main_correct
+    print(
+        f"Change in correct rows (higher == better): {correct_change:+d} ({100*correct_change/main_correct:+.1f}%)"
```
Comment (siddharth-krishna) [inline utils/run_benchmarks.py:357]: These changes are great, thanks!

**Review by siddharth-krishna — APPROVED** [review]
Looks good! Go ahead with the merge when ready. :shipit:

File: README.md (hunk 6 above)
Comment (SamRWest) [reply README.md:137]: thanks, that was a merge error.  fixed

File: utils/run_benchmarks.py (hunk 7 above)
Comment (SamRWest) [reply utils/run_benchmarks.py:357]: No worries :)
I wasn't sure how to interpret these metrics at first.

**Pull request merged**